GSPath: Implement sliced object downloads

Sliced downloads are beneficial in cases of downloading a small number of large files. GCP documentation here with example code using the cloud storage python client:

https://cloud.google.com/storage/docs/sliced-object-downloads

Unsure if this is within scope of cloudpathlib but such a feature would be useful.

Sure thing:

```
tests/test_gs_specific.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cloudpathlib/cloudpath.py:893: in download_to
    return self.client._download_file(self, destination)
cloudpathlib/gs/gsclient.py:129: in _download_file
    transfer_manager.download_chunks_concurrently(
venv/lib/python3.11/site-packages/google/cloud/storage/transfer_manager.py:895: in download_chunks_concurrently
    maybe_pickled_blob = _pickle_client(blob) if needs_pickling else blob
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <tests.mock_clients.mock_gs.MockBlob object at 0x11d96e7d0>

    def _pickle_client(obj):
        """Pickle a Client or an object that owns a Client (like a Blob)"""
    
        # We need a custom pickler to process Client objects, which are attached to
        # Buckets (and therefore to Blobs in turn). Unfortunately, the Python
        # multiprocessing library doesn't seem to have a good way to use a custom
        # pickler, and using copyreg will mutate global state and affect code
        # outside of the client library. Instead, we'll pre-pickle the object and
        # pass the bytestring in.
        f = io.BytesIO()
        p = pickle.Pickler(f)
        p.dispatch_table = copyreg.dispatch_table.copy()
        p.dispatch_table[Client] = _reduce_client
>       p.dump(obj)
E       AttributeError: Can't pickle local object 'mocked_client_class_factory.<locals>.MockClient'

venv/lib/python3.11/site-packages/google/cloud/storage/transfer_manager.py:1318: AttributeError
```

I managed to follow the thread a little. `storage_manager` needs to pickle the client when using `worker_type="process"` as noted e.g. [here](https://github.com/googleapis/python-storage/blob/828f529b0870d962ecaaf3b06765ca7b898dc244/google/cloud/storage/transfer_manager.py#L323). The test passes if you pass `worker_type="thread"` via `download_chunks_concurrently_kwargs`.

--- a/HISTORY.md
+++ b/HISTORY.md
@@ -3,6 +3,7 @@
 ## Unreleased
 - Fix `S3Client` cleanup via `Client.__del__` when `S3Client` encounters an exception during initialization. (Issue [#372](https://github.com/drivendataorg/cloudpathlib/issues/372), PR [#373](https://github.com/drivendataorg/cloudpathlib/pull/373))
 - Skip mtime checks during upload when force_overwrite_to_cloud is set to improve upload performance. (Issue [#379](https://github.com/drivendataorg/cloudpathlib/issues/379), PR [#380](https://github.com/drivendataorg/cloudpathlib/pull/380), thanks to [@Gilthans](https://github.com/Gilthans))
+- Implement sliced downloads in GSClient. (Issue [#387](https://github.com/drivendataorg/cloudpathlib/issues/387), PR [#389](https://github.com/drivendataorg/cloudpathlib/pull/389))
 
 ## v0.16.0 (2023-10-09)
  - Add "CloudPath" as return type on `__init__` for mypy issues. ([Issue #179](https://github.com/drivendataorg/cloudpathlib/issues/179), [PR #342](https://github.com/drivendataorg/cloudpathlib/pull/342))

--- a/cloudpathlib/gs/gsclient.py
+++ b/cloudpathlib/gs/gsclient.py
@@ -1,8 +1,8 @@
-from datetime import datetime
 import mimetypes
 import os
+from datetime import datetime
 from pathlib import Path, PurePosixPath
-from typing import Any, Callable, Dict, Iterable, Optional, TYPE_CHECKING, Tuple, Union
+from typing import TYPE_CHECKING, Any, Callable, Dict, Iterable, Optional, Tuple, Union
 
 from ..client import Client, register_client_class
 from ..cloudpath import implementation_registry
@@ -16,6 +16,7 @@ try:
     from google.api_core.exceptions import NotFound
     from google.auth.exceptions import DefaultCredentialsError
     from google.cloud.storage import Client as StorageClient
+    from google.cloud.storage import transfer_manager
 
 
 except ModuleNotFoundError:
@@ -39,6 +40,7 @@ class GSClient(Client):
         file_cache_mode: Optional[Union[str, FileCacheMode]] = None,
         local_cache_dir: Optional[Union[str, os.PathLike]] = None,
         content_type_method: Optional[Callable] = mimetypes.guess_type,
+        download_chunks_concurrently_kwargs: Optional[Dict[str, Any]] = None,
     ):
         """Class constructor. Sets up a [`Storage
         Client`](https://googleapis.dev/python/storage/latest/client.html).
@@ -76,6 +78,9 @@ class GSClient(Client):
                 the `CLOUDPATHLIB_LOCAL_CACHE_DIR` environment variable.
             content_type_method (Optional[Callable]): Function to call to guess media type (mimetype) when
                 writing a file to the cloud. Defaults to `mimetypes.guess_type`. Must return a tuple (content type, content encoding).
+            download_chunks_concurrently_kwargs (Optional[Dict[str, Any]]): Keyword arguments to pass to
+                [`download_chunks_concurrently`](https://cloud.google.com/python/docs/reference/storage/latest/google.cloud.storage.transfer_manager#google_cloud_storage_transfer_manager_download_chunks_concurrently)
+                for sliced parallel downloads.
         """
         if application_credentials is None:
             application_credentials = os.getenv("GOOGLE_APPLICATION_CREDENTIALS")
@@ -91,6 +96,8 @@ class GSClient(Client):
                 self.client = StorageClient()
             except DefaultCredentialsError:
                 self.client = StorageClient.create_anonymous_client()
+
+        self.download_chunks_concurrently_kwargs = download_chunks_concurrently_kwargs
 
         super().__init__(
             local_cache_dir=local_cache_dir,
@@ -118,7 +125,13 @@ class GSClient(Client):
 
         local_path = Path(local_path)
 
-        blob.download_to_filename(local_path)
+        if self.download_chunks_concurrently_kwargs is not None:
+            transfer_manager.download_chunks_concurrently(
+                blob, local_path, **self.download_chunks_concurrently_kwargs
+            )
+        else:
+            blob.download_to_filename(local_path)
+
         return local_path
 
     def _is_file_or_dir(self, cloud_path: GSPath) -> Optional[str]:

--- a/cloudpathlib/gs/gsclient.py
+++ b/cloudpathlib/gs/gsclient.py
@@ -1,8 +1,8 @@
+from datetime import datetime
 import mimetypes
 import os
-from datetime import datetime
 from pathlib import Path, PurePosixPath
-from typing import TYPE_CHECKING, Any, Callable, Dict, Iterable, Optional, Tuple, Union
+from typing import Any, Callable, Dict, Iterable, Optional, TYPE_CHECKING, Tuple, Union
 
 from ..client import Client, register_client_class
 from ..cloudpath import implementation_registry

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -30,6 +30,7 @@ from .mock_clients.mock_azureblob import mocked_client_class_factory, DEFAULT_CO
 from .mock_clients.mock_gs import (
     mocked_client_class_factory as mocked_gsclient_class_factory,
     DEFAULT_GS_BUCKET_NAME,
+    MockTransferManager,
 )
 from .mock_clients.mock_s3 import mocked_session_class_factory, DEFAULT_S3_BUCKET_NAME
 
@@ -183,6 +184,11 @@ def gs_rig(request, monkeypatch, assets_dir):
             cloudpathlib.gs.gsclient,
             "StorageClient",
             mocked_gsclient_class_factory(test_dir),
+        )
+        monkeypatch.setattr(
+            cloudpathlib.gs.gsclient,
+            "transfer_manager",
+            MockTransferManager,
         )
 
     rig = CloudProviderTestRig(

--- a/tests/mock_clients/mock_gs.py
+++ b/tests/mock_clients/mock_gs.py
@@ -168,3 +168,19 @@ class MockHTTPIterator:
     @property
     def prefixes(self):
         return self.sub_directories
+
+
+class MockTransferManager:
+    @staticmethod
+    def download_chunks_concurrently(
+        blob,
+        filename,
+        chunk_size=32 * 1024 * 1024,
+        download_kwargs=None,
+        deadline=None,
+        worker_type="process",
+        max_workers=8,
+        *,
+        crc32c_checksum=True,
+    ):
+        blob.download_to_filename(filename)

--- a/tests/mock_clients/mock_gs.py
+++ b/tests/mock_clients/mock_gs.py
@@ -1,6 +1,6 @@
+import shutil
 from datetime import datetime
 from pathlib import Path, PurePosixPath
-import shutil
 from tempfile import TemporaryDirectory
 
 from google.api_core.exceptions import NotFound
@@ -64,6 +64,21 @@ class MockBlob:
     def patch(self):
         if "updated" in self.metadata:
             (self.bucket / self.name).touch()
+
+    def reload(
+        self,
+        client=None,
+        projection="noAcl",
+        if_etag_match=None,
+        if_etag_not_match=None,
+        if_generation_match=None,
+        if_generation_not_match=None,
+        if_metageneration_match=None,
+        if_metageneration_not_match=None,
+        timeout=None,
+        retry=None,
+    ):
+        pass
 
     def upload_from_filename(self, filename, content_type=None):
         data = Path(filename).read_bytes()

--- a/tests/mock_clients/mock_gs.py
+++ b/tests/mock_clients/mock_gs.py
@@ -1,6 +1,6 @@
-import shutil
 from datetime import datetime
 from pathlib import Path, PurePosixPath
+import shutil
 from tempfile import TemporaryDirectory
 
 from google.api_core.exceptions import NotFound

--- a/tests/test_gs_specific.py
+++ b/tests/test_gs_specific.py
@@ -15,8 +15,9 @@ def test_gspath_properties(path_class):
     assert p2.bucket == "mybucket"
 
 
-def test_concurrent_download(gs_rig, tmp_path):
-    client = gs_rig.client_class(download_chunks_concurrently_kwargs={})
+@pytest.mark.parametrize("worker_type", ["process", "thread"])
+def test_concurrent_download(gs_rig, tmp_path, worker_type):
+    client = gs_rig.client_class(download_chunks_concurrently_kwargs={"worker_type": worker_type})
     p = gs_rig.create_cloud_path("dir_0/file0_0.txt", client=client)
     dl_dir = tmp_path
     assert not (dl_dir / p.name).exists()

--- a/tests/test_gs_specific.py
+++ b/tests/test_gs_specific.py
@@ -1,7 +1,7 @@
 import pytest
 
-from cloudpathlib import GSClient, GSPath
-from cloudpathlib.local import LocalGSClient, LocalGSPath
+from cloudpathlib import GSPath
+from cloudpathlib.local import LocalGSPath
 
 
 @pytest.mark.parametrize("path_class", [GSPath, LocalGSPath])

--- a/tests/test_gs_specific.py
+++ b/tests/test_gs_specific.py
@@ -1,7 +1,7 @@
 import pytest
 
-from cloudpathlib import GSPath
-from cloudpathlib.local import LocalGSPath
+from cloudpathlib import GSClient, GSPath
+from cloudpathlib.local import LocalGSClient, LocalGSPath
 
 
 @pytest.mark.parametrize("path_class", [GSPath, LocalGSPath])
@@ -13,3 +13,12 @@ def test_gspath_properties(path_class):
     p2 = path_class("gs://mybucket/")
     assert p2.blob == ""
     assert p2.bucket == "mybucket"
+
+
+def test_concurrent_download(gs_rig, tmp_path):
+    client = gs_rig.client_class(download_chunks_concurrently_kwargs={})
+    p = gs_rig.create_cloud_path("dir_0/file0_0.txt", client=client)
+    dl_dir = tmp_path
+    assert not (dl_dir / p.name).exists()
+    p.download_to(dl_dir)
+    assert (dl_dir / p.name).is_file()

